frame/nis/src/lib.rs

```diff
@@ -0,0 +1,937 @@
+			/// The proportion of the effective total issuance which the receipt represents.
+			proportion: Perquintill,
+			/// The amount of funds which were debited from the owner.
+			amount: BalanceOf<T>,
+		},
+		/// An receipt has been (at least partially) thawed.
+		Thawed {
+			/// The identity of the receipt.
+			index: ReceiptIndex,
+			/// The owner.
+			who: T::AccountId,
+			/// The proportion of the effective total issuance by which the owner was debited.
+			proportion: Perquintill,
+			/// The amount by which the owner was credited.
+			amount: BalanceOf<T>,
+			/// If `true` then the receipt is done.
+			dropped: bool,
+		},
+		/// An automatic funding of the deficit was made.
+		Funded { deficit: BalanceOf<T> },
+		/// A receipt was transfered.
+		Transferred { from: T::AccountId, to: T::AccountId, index: ReceiptIndex },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The duration of the bid is less than one.
+		DurationTooSmall,
+		/// The duration is the bid is greater than the number of queues.
+		DurationTooBig,
+		/// The amount of the bid is less than the minimum allowed.
+		AmountTooSmall,
+		/// The queue for the bid's duration is full and the amount bid is too low to get in
+		/// through replacing an existing bid.
+		BidTooLow,
+		/// Bond index is unknown.
+		Unknown,
+		/// Not the owner of the receipt.
+		NotOwner,
+		/// Bond not yet at expiry date.
+		NotExpired,
+		/// The given bid for retraction is not found.
+		NotFound,
+		/// The portion supplied is beyond the value of the receipt.
+		TooMuch,
+		/// Not enough funds are held to pay out.
+		Unfunded,
+		/// There are enough funds for what is required.
+		Funded,
+		/// The thaw throttle has been reached for this period.
+		Throttled,
+		/// The operation would result in a receipt worth an insignficant value.
+		MakesDust,
+	}
+
+	pub(crate) struct WeightCounter {
+		pub(crate) used: Weight,
+		pub(crate) limit: Weight,
+	}
+	impl WeightCounter {
+		#[allow(dead_code)]
+		pub(crate) fn unlimited() -> Self {
+			WeightCounter { used: Weight::zero(), limit: Weight::max_value() }
+		}
+		fn check_accrue(&mut self, w: Weight) -> bool {
+			let test = self.used.saturating_add(w);
+			if test.any_gt(self.limit) {
+				false
+			} else {
+				self.used = test;
+				true
+			}
+		}
+		fn can_accrue(&mut self, w: Weight) -> bool {
+			self.used.saturating_add(w).all_lte(self.limit)
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			let mut weight_counter =
+				WeightCounter { used: Weight::zero(), limit: T::MaxIntakeWeight::get() };
+			if T::IntakePeriod::get().is_zero() || (n % T::IntakePeriod::get()).is_zero() {
+				if weight_counter.check_accrue(T::WeightInfo::process_queues()) {
+					Self::process_queues(
+						T::Target::get(),
+						T::QueueCount::get(),
+						u32::max_value(),
+						&mut weight_counter,
+					)
+				}
+			}
+			weight_counter.used
+		}
```

Comment:
```suggestion
		}
		
		fn integrity_test() {
			assert!(!T::IntakePeriod::get().is_zero());
		}
```

Reply:
latter because of: 

```
let queue_full = q.len() == T::MaxQueueLen::get() as usize;
ensure!(!queue_full || q[0].amount < amount, Error::<T>::BidTooLow);
```